File: Algrithms-Level-1/day-5/solution1.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode *middleNode(struct ListNode *head)
{
    struct ListNode *middle = head;
    uint8_t i = 0;
    while (head != NULL)
    {
        if (i % 2 == 1)
            middle = middle->next;
        head = head->next;
        i++;
    }
    return middle;
}

File: Algrithms-Level-1/day-5/solution1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        def recursive(node):
            if node.next == None:
                return [node]

            result = [node]
            result += recursive(node.next)
            return result

        final_list = recursive(head)
        return final_list[len(final_list)//2]
